x-pack/plugins/security/server/session_management/session_index.ts

```diff
@@ -0,0 +1,386 @@
+    }
+
+    const { sid, ...sessionValueToStore } = sessionValue;
+    try {
+      const {
+        _primary_term: primaryTerm,
+        _seq_no: sequenceNumber,
+      } = await this.#options.clusterClient.callAsInternalUser('create', {
+        id: sid,
+        // We cannot control whether index is created automatically during this operation or not.
+        // But we can reduce probability of getting into a weird state when session is being created
+        // while session index is missing for some reason. This way we'll recreate index with a
+        // proper name and alias. But this will only work if we still have a proper index template.
+        index: SESSION_INDEX_NAME,
+        body: sessionValueToStore,
+        refresh: 'wait_for',
+      });
+
+      return { ...sessionValue, metadata: { primaryTerm, sequenceNumber } } as SessionIndexValue;
+    } catch (err) {
+      this.#options.logger.error(`Failed to create session value: ${err.message}`);
+      throw err;
+    }
+  }
+
+  /**
+   * Re-indexes updated session value.
+   * @param sessionValue Session index value.
+   */
+  async update(sessionValue: Readonly<SessionIndexValue>) {
+    const { sid, metadata, ...sessionValueToStore } = sessionValue;
+    try {
+      const response = await this.#options.clusterClient.callAsInternalUser('index', {
+        id: sid,
+        index: SESSION_INDEX_ALIAS,
+        body: sessionValueToStore,
+        ifSeqNo: metadata.sequenceNumber,
+        ifPrimaryTerm: metadata.primaryTerm,
+        refresh: 'wait_for',
+        ignore: [409],
+      });
+
+      // We don't want to override changes that were made after we fetched session value or
+      // re-create it if has been deleted already. If we detect such a case we discard changes and
+      // return latest copy of the session value instead or `null` if doesn't exist anymore.
+      const sessionIndexValueUpdateConflict = response.status === 409;
+      if (sessionIndexValueUpdateConflict) {
+        this.#options.logger.debug(
+          'Cannot update session value due to conflict, session either do not exist or was already updated.'
+        );
+        return await this.get(sid);
+      }
+
+      return {
+        ...sessionValue,
+        metadata: { primaryTerm: response._primary_term, sequenceNumber: response._seq_no },
+      } as SessionIndexValue;
+    } catch (err) {
+      this.#options.logger.error(`Failed to update session value: ${err.message}`);
+      throw err;
+    }
+  }
+
+  /**
+   * Clears session value with the specified ID. May trigger a removal of other outdated session
+   * values.
+   * @param sid Session ID to clear.
+   */
+  async clear(sid: string) {
+    try {
+      const now = Date.now();
+
+      // Always try to delete session with the specified ID and with expired lifespan (even if it's
+      // not configured right now).
+      // QUESTION: CAN WE SAY THAT ALL TENANTS SHOULD HAVE THE SAME SESSION TIMEOUTS?
```

Comment:
__note:__ that's the question :slightly_smiling_face: But it sounds like something we can declare as a requirement. What do you think?

Reply:
I'm afraid that we may end up with too many dangling sessions that we can not confidently mark as invalid... Do you know if we have any unique identifier of tenant we can rely on?

Reply:
Okay, with your suggestion regarding using "tenant" identifier (kibana index name) we don't need this restriction, thanks! I know you mentioned this a while ago, but it somehow slipped under my radar :/